{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"20190410-Roche-LN-pt_table.csv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Region</th>\n",
       "      <th>Chain</th>\n",
       "      <th>MID_no</th>\n",
       "      <th>Primer</th>\n",
       "      <th>MID</th>\n",
       "      <th>Patient</th>\n",
       "      <th>Sample_Project</th>\n",
       "      <th>Reads</th>\n",
       "      <th>Species</th>\n",
       "      <th>Sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>run169</td>\n",
       "      <td>region1</td>\n",
       "      <td>BCRh</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGTCAGAGGC</td>\n",
       "      <td>LN40</td>\n",
       "      <td>LN-study</td>\n",
       "      <td>2500</td>\n",
       "      <td>human</td>\n",
       "      <td>10-IgA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>run169</td>\n",
       "      <td>region2</td>\n",
       "      <td>BCRh</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGTCAGAGGC</td>\n",
       "      <td>LN75</td>\n",
       "      <td>LN-study</td>\n",
       "      <td>2500</td>\n",
       "      <td>human</td>\n",
       "      <td>10-IgA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>run169</td>\n",
       "      <td>region1</td>\n",
       "      <td>BCRh</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGTCATGTCT</td>\n",
       "      <td>LN40</td>\n",
       "      <td>LN-study</td>\n",
       "      <td>2500</td>\n",
       "      <td>human</td>\n",
       "      <td>10-IgD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>run169</td>\n",
       "      <td>region2</td>\n",
       "      <td>BCRh</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGTCATGTCT</td>\n",
       "      <td>LN75</td>\n",
       "      <td>LN-study</td>\n",
       "      <td>2500</td>\n",
       "      <td>human</td>\n",
       "      <td>10-IgD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>run169</td>\n",
       "      <td>region1</td>\n",
       "      <td>BCRh</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGTCAGGGGA</td>\n",
       "      <td>LN40</td>\n",
       "      <td>LN-study</td>\n",
       "      <td>2500</td>\n",
       "      <td>human</td>\n",
       "      <td>10-IgG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Run   Region Chain  MID_no Primer         MID Patient Sample_Project  \\\n",
       "0  run169  region1  BCRh      10    NaN  AGTCAGAGGC    LN40       LN-study   \n",
       "1  run169  region2  BCRh      10    NaN  AGTCAGAGGC    LN75       LN-study   \n",
       "2  run169  region1  BCRh      10    NaN  AGTCATGTCT    LN40       LN-study   \n",
       "3  run169  region2  BCRh      10    NaN  AGTCATGTCT    LN75       LN-study   \n",
       "4  run169  region1  BCRh      10    NaN  AGTCAGGGGA    LN40       LN-study   \n",
       "\n",
       "   Reads Species  Sample  \n",
       "0   2500   human  10-IgA  \n",
       "1   2500   human  10-IgA  \n",
       "2   2500   human  10-IgD  \n",
       "3   2500   human  10-IgD  \n",
       "4   2500   human  10-IgG  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "ig = df[\"Sample\"].tolist()\n",
    "ig_new = list()\n",
    "for i in range(len(ig)):\n",
    "    if \"-\" in str(ig[i]):\n",
    "        ig_new.append(ig[i].split(\"-\")[1])\n",
    "    else:\n",
    "        ig_new.append('')\n",
    "df[\"Sample\"] = ig_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Region</th>\n",
       "      <th>Chain</th>\n",
       "      <th>MID_no</th>\n",
       "      <th>Primer</th>\n",
       "      <th>MID</th>\n",
       "      <th>Patient</th>\n",
       "      <th>Sample_Project</th>\n",
       "      <th>Reads</th>\n",
       "      <th>Species</th>\n",
       "      <th>Sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>run169</td>\n",
       "      <td>region1</td>\n",
       "      <td>BCRh</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGTCAGAGGC</td>\n",
       "      <td>LN40</td>\n",
       "      <td>LN-study</td>\n",
       "      <td>2500</td>\n",
       "      <td>human</td>\n",
       "      <td>IgA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>run169</td>\n",
       "      <td>region2</td>\n",
       "      <td>BCRh</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGTCAGAGGC</td>\n",
       "      <td>LN75</td>\n",
       "      <td>LN-study</td>\n",
       "      <td>2500</td>\n",
       "      <td>human</td>\n",
       "      <td>IgA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>run169</td>\n",
       "      <td>region1</td>\n",
       "      <td>BCRh</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGTCATGTCT</td>\n",
       "      <td>LN40</td>\n",
       "      <td>LN-study</td>\n",
       "      <td>2500</td>\n",
       "      <td>human</td>\n",
       "      <td>IgD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>run169</td>\n",
       "      <td>region2</td>\n",
       "      <td>BCRh</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGTCATGTCT</td>\n",
       "      <td>LN75</td>\n",
       "      <td>LN-study</td>\n",
       "      <td>2500</td>\n",
       "      <td>human</td>\n",
       "      <td>IgD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>run169</td>\n",
       "      <td>region1</td>\n",
       "      <td>BCRh</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGTCAGGGGA</td>\n",
       "      <td>LN40</td>\n",
       "      <td>LN-study</td>\n",
       "      <td>2500</td>\n",
       "      <td>human</td>\n",
       "      <td>IgG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Run   Region Chain  MID_no Primer         MID Patient Sample_Project  \\\n",
       "0  run169  region1  BCRh      10    NaN  AGTCAGAGGC    LN40       LN-study   \n",
       "1  run169  region2  BCRh      10    NaN  AGTCAGAGGC    LN75       LN-study   \n",
       "2  run169  region1  BCRh      10    NaN  AGTCATGTCT    LN40       LN-study   \n",
       "3  run169  region2  BCRh      10    NaN  AGTCATGTCT    LN75       LN-study   \n",
       "4  run169  region1  BCRh      10    NaN  AGTCAGGGGA    LN40       LN-study   \n",
       "\n",
       "   Reads Species Sample  \n",
       "0   2500   human    IgA  \n",
       "1   2500   human    IgA  \n",
       "2   2500   human    IgD  \n",
       "3   2500   human    IgD  \n",
       "4   2500   human    IgG  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "patient = df[\"Patient\"].tolist()\n",
    "subtype = df[\"Sample\"].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_name = list()\n",
    "for i in range(len(patient)):\n",
    "    if subtype[i] != '':\n",
    "        sample_name.append(patient[i] + \"-\" + subtype[i])\n",
    "    else:\n",
    "        sample_name.append(patient[i])\n",
    "df[\"Sample_Name\"] = sample_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "465"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"20190410-Roche-LN-pt_table-reformatted.csv\", index=False, sep=\";\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## List split files and write linux commands to rename them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Region</th>\n",
       "      <th>MID</th>\n",
       "      <th>File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>run162</td>\n",
       "      <td>region2</td>\n",
       "      <td>ACACTGCTCA</td>\n",
       "      <td>run162-r2_L001.assembled-ACACTGCTCA.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>run162</td>\n",
       "      <td>region2</td>\n",
       "      <td>ACAGGGCTCA</td>\n",
       "      <td>run162-r2_L001.assembled-ACAGGGCTCA.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>run162</td>\n",
       "      <td>region2</td>\n",
       "      <td>ACATGAGCTC</td>\n",
       "      <td>run162-r2_L001.assembled-ACATGAGCTC.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>run162</td>\n",
       "      <td>region2</td>\n",
       "      <td>ACTAGGGCTC</td>\n",
       "      <td>run162-r2_L001.assembled-ACTAGGGCTC.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>run162</td>\n",
       "      <td>region2</td>\n",
       "      <td>AGAACGGCTC</td>\n",
       "      <td>run162-r2_L001.assembled-AGAACGGCTC.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Run   Region         MID                                          File\n",
       "0  run162  region2  ACACTGCTCA  run162-r2_L001.assembled-ACACTGCTCA.fastq.gz\n",
       "1  run162  region2  ACAGGGCTCA  run162-r2_L001.assembled-ACAGGGCTCA.fastq.gz\n",
       "2  run162  region2  ACATGAGCTC  run162-r2_L001.assembled-ACATGAGCTC.fastq.gz\n",
       "3  run162  region2  ACTAGGGCTC  run162-r2_L001.assembled-ACTAGGGCTC.fastq.gz\n",
       "4  run162  region2  AGAACGGCTC  run162-r2_L001.assembled-AGAACGGCTC.fastq.gz"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mydir = 'split'\n",
    "myfiles = list()\n",
    "myrun = list()\n",
    "myregion = list()\n",
    "mymid = list()\n",
    "for f in os.listdir(mydir):\n",
    "    if f.endswith('.fastq.gz'):\n",
    "        (run, region, mid) = f.split(\"-\")\n",
    "        region = region.split(\"_\")[0]\n",
    "        region = region.replace(\"r\", \"region\")\n",
    "        mid = mid.replace(\".fastq.gz\", \"\")\n",
    "        \n",
    "        myfiles.append(f)\n",
    "        myrun.append(run)\n",
    "        myregion.append(region)\n",
    "        mymid.append(mid)\n",
    "d = {'Run': myrun, 'Region': myregion, 'MID': mymid, 'File': myfiles}\n",
    "df_files = pd.DataFrame.from_dict(d)\n",
    "df_files.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine with the sample information\n",
    "df_merged = pd.merge(df, df_files, how='left', on=['Run', 'Region', 'MID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create new file name\n",
    "sample_name = df_merged['Sample_Name'].tolist()\n",
    "file_name = df_merged['File'].tolist()\n",
    "new_name = list()\n",
    "for i in range(len(sample_name)):\n",
    "    new_name.append(str(sample_name[i]) + \"-\" + str(file_name[i]))\n",
    "df_merged['File_Renamed'] = new_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Region</th>\n",
       "      <th>Chain</th>\n",
       "      <th>MID_no</th>\n",
       "      <th>Primer</th>\n",
       "      <th>MID</th>\n",
       "      <th>Patient</th>\n",
       "      <th>Sample_Project</th>\n",
       "      <th>Reads</th>\n",
       "      <th>Species</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>File</th>\n",
       "      <th>File_Renamed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>run169</td>\n",
       "      <td>region1</td>\n",
       "      <td>BCRh</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGTCAGAGGC</td>\n",
       "      <td>LN40</td>\n",
       "      <td>LN-study</td>\n",
       "      <td>2500</td>\n",
       "      <td>human</td>\n",
       "      <td>IgA</td>\n",
       "      <td>LN40-IgA</td>\n",
       "      <td>run169-r1_L001.assembled-AGTCAGAGGC.fastq.gz</td>\n",
       "      <td>LN40-IgA-run169-r1_L001.assembled-AGTCAGAGGC.f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>run169</td>\n",
       "      <td>region2</td>\n",
       "      <td>BCRh</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGTCAGAGGC</td>\n",
       "      <td>LN75</td>\n",
       "      <td>LN-study</td>\n",
       "      <td>2500</td>\n",
       "      <td>human</td>\n",
       "      <td>IgA</td>\n",
       "      <td>LN75-IgA</td>\n",
       "      <td>run169-r2_L001.assembled-AGTCAGAGGC.fastq.gz</td>\n",
       "      <td>LN75-IgA-run169-r2_L001.assembled-AGTCAGAGGC.f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>run169</td>\n",
       "      <td>region1</td>\n",
       "      <td>BCRh</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGTCATGTCT</td>\n",
       "      <td>LN40</td>\n",
       "      <td>LN-study</td>\n",
       "      <td>2500</td>\n",
       "      <td>human</td>\n",
       "      <td>IgD</td>\n",
       "      <td>LN40-IgD</td>\n",
       "      <td>run169-r1_L001.assembled-AGTCATGTCT.fastq.gz</td>\n",
       "      <td>LN40-IgD-run169-r1_L001.assembled-AGTCATGTCT.f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>run169</td>\n",
       "      <td>region2</td>\n",
       "      <td>BCRh</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGTCATGTCT</td>\n",
       "      <td>LN75</td>\n",
       "      <td>LN-study</td>\n",
       "      <td>2500</td>\n",
       "      <td>human</td>\n",
       "      <td>IgD</td>\n",
       "      <td>LN75-IgD</td>\n",
       "      <td>run169-r2_L001.assembled-AGTCATGTCT.fastq.gz</td>\n",
       "      <td>LN75-IgD-run169-r2_L001.assembled-AGTCATGTCT.f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>run169</td>\n",
       "      <td>region1</td>\n",
       "      <td>BCRh</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGTCAGGGGA</td>\n",
       "      <td>LN40</td>\n",
       "      <td>LN-study</td>\n",
       "      <td>2500</td>\n",
       "      <td>human</td>\n",
       "      <td>IgG</td>\n",
       "      <td>LN40-IgG</td>\n",
       "      <td>run169-r1_L001.assembled-AGTCAGGGGA.fastq.gz</td>\n",
       "      <td>LN40-IgG-run169-r1_L001.assembled-AGTCAGGGGA.f...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Run   Region Chain  MID_no Primer         MID Patient Sample_Project  \\\n",
       "0  run169  region1  BCRh      10    NaN  AGTCAGAGGC    LN40       LN-study   \n",
       "1  run169  region2  BCRh      10    NaN  AGTCAGAGGC    LN75       LN-study   \n",
       "2  run169  region1  BCRh      10    NaN  AGTCATGTCT    LN40       LN-study   \n",
       "3  run169  region2  BCRh      10    NaN  AGTCATGTCT    LN75       LN-study   \n",
       "4  run169  region1  BCRh      10    NaN  AGTCAGGGGA    LN40       LN-study   \n",
       "\n",
       "   Reads Species Sample Sample_Name  \\\n",
       "0   2500   human    IgA    LN40-IgA   \n",
       "1   2500   human    IgA    LN75-IgA   \n",
       "2   2500   human    IgD    LN40-IgD   \n",
       "3   2500   human    IgD    LN75-IgD   \n",
       "4   2500   human    IgG    LN40-IgG   \n",
       "\n",
       "                                           File  \\\n",
       "0  run169-r1_L001.assembled-AGTCAGAGGC.fastq.gz   \n",
       "1  run169-r2_L001.assembled-AGTCAGAGGC.fastq.gz   \n",
       "2  run169-r1_L001.assembled-AGTCATGTCT.fastq.gz   \n",
       "3  run169-r2_L001.assembled-AGTCATGTCT.fastq.gz   \n",
       "4  run169-r1_L001.assembled-AGTCAGGGGA.fastq.gz   \n",
       "\n",
       "                                        File_Renamed  \n",
       "0  LN40-IgA-run169-r1_L001.assembled-AGTCAGAGGC.f...  \n",
       "1  LN75-IgA-run169-r2_L001.assembled-AGTCAGAGGC.f...  \n",
       "2  LN40-IgD-run169-r1_L001.assembled-AGTCATGTCT.f...  \n",
       "3  LN75-IgD-run169-r2_L001.assembled-AGTCATGTCT.f...  \n",
       "4  LN40-IgG-run169-r1_L001.assembled-AGTCAGGGGA.f...  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote tmp-rename.sh to disk\n"
     ]
    }
   ],
   "source": [
    "fhOut = open(\"tmp-rename.sh\", \"w\")\n",
    "for i in range(len(df_merged)):\n",
    "    if str(df_merged['File'][i]).endswith(\".fastq.gz\"):\n",
    "        print(\"mv\", df_merged['File'][i], df_merged['File_Renamed'][i], file=fhOut)\n",
    "fhOut.close()\n",
    "print(\"Wrote tmp-rename.sh to disk\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
